plugins/modules/win_environment.ps1

```diff
@@ -1,61 +1,124 @@
+        [System.EnvironmentVariableTarget]
+        $Level ,
+
+        [Parameter(Mandatory=$true, ValueFromPipelineByPropertyName=$true)]
+        [Alias('Key')]
+        [String]
+        $Name ,
+
+        [Parameter(ValueFromPipelineByPropertyName=$true)]
+        [String]
+        $Value,
+
+        [Parameter()]
+        [ValidateSet('present', 'absent')]
+        [String]
+        $State
+    )
+
+    Process {
+        $before_value = [Environment]::GetEnvironmentVariable($name, $Level)
+
+        $ret = @{
+            changed = $false
+            after = $Value
+            before = $before_value
+        }
+
+        if ($State -eq "present" -and $before_value -ne $Value) {
+            if ($PSCmdlet.ShouldProcess($Name, 'Set environment variable')) {
+                [Environment]::SetEnvironmentVariable($Name, $Value, $Level)
+            }
+            $ret.changed = $true
+        } elseif ($State -eq "absent" -and $null -ne $before_value) {
+            if ($PSCmdlet.ShouldProcess($Name, 'Remove environment variable')) {
+                [Environment]::SetEnvironmentVariable($Name, $null, $Level)
+            }
+            $ret.changed = $true
+        }
+
+        $ret
+    }
+}
+
+$module.Result.values = @{}
+
 $level = $module.Params.level
 $state = $module.Params.state
-$value = $module.Params.value
-
-$before_value = [Environment]::GetEnvironmentVariable($name, $level)
-$module.Result.before_value = $before_value
-$module.Result.value = $value
-
-# When removing environment, set value to $null if set
-if ($state -eq "absent" -and $value) {
-    $module.Warn("When removing environment variable '$name' it should not have a value '$value' set")
-    $value = $null
-} elseif ($state -eq "present" -and (-not $value)) {
-    $module.FailJson("When state=present, value must be defined and not an empty string, if you wish to remove the envvar, set state=absent")
+
+$envvars = if ($module.Params.variables) {
+    $module.Params.variables
+}
+else {
+    @{
+        $module.Params.name = $module.Params.value
+    }
 }
 
 $module.Diff.before = @{ $level = @{} }
-if ($before_value) {
-    $module.Diff.before.$level.$name = $before_value
-}
 $module.Diff.after = @{ $level = @{} }
-if ($value) {
-    $module.Diff.after.$level.$name = $value
-}
 
-if ($state -eq "present" -and $before_value -ne $value) {
-    if (-not $module.CheckMode) {
-        [Environment]::SetEnvironmentVariable($name, $value, $level)
+foreach ($kv in $envvars.GetEnumerator()) {
+    $name = $kv.Key
+    $value = $kv.Value
+
+    # When removing environment, set value to $null if set
+    if ($state -eq "absent" -and $value) {
+        $module.Warn("When removing environment variable '$name' it should not have a value '$value' set")
+        $value = $null
+    } elseif ($state -eq "present" -and (-not $value)) {
+        $module.FailJson("When state=present, value must be defined and not an empty string, if you wish to remove the envvar, set state=absent")
```

Comment:
@jborean93 this is the part (from the original, just moved in this PR) that catches the case of `state=present` and `value=<null/empty>`